maas/resource_maas_network_interface_bond.go

```diff
@@ -0,0 +1,310 @@
+	}
+	if _, ok := p["bond_updelay"]; ok {
+		d.Set("bond_updelay", int64(p["bond_updelay"].(float64)))
+	}
+	if _, ok := p["bond_xmit_hash_policy"]; ok {
+		d.Set("bond_xmit_hash_policy", p["bond_xmit_hash_policy"].(string))
+	}
+
+	tfState := map[string]interface{}{
+		"name":        networkInterface.Name,
+		"parents":     networkInterface.Parents,
+		"accept_ra":   networkInterface.AcceptRA,
+		"autoconf":    networkInterface.Autoconf,
+		"mac_address": networkInterface.MACAddress,
+		"mtu":         networkInterface.EffectiveMTU,
+		"tags":        networkInterface.Tags,
+		"vlan":        fmt.Sprintf("%v", networkInterface.VLAN.ID),
+	}
+	if err := setTerraformState(d, tfState); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceMaasNetworkInterfaceBondUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*client.Client)
+
+	machine, err := getMachine(client, d.Get("machine").(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	p, err := findBondParentsID(client, machine.SystemID, d.Get("parents").(*schema.Set).List())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	params := getNetworkInterfaceBondParams(d, p)
+	_, err = client.NetworkInterface.Update(machine.SystemID, id, params)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceMaasNetworkInterfaceBondRead(ctx, d, m)
+}
+
+func resourceMaasNetworkInterfaceBondDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*client.Client)
+
+	machine, err := getMachine(client, d.Get("machine").(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if err := client.NetworkInterface.Delete(machine.SystemID, id); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func getNetworkInterfaceBondParams(d *schema.ResourceData, parentIDs []int) *entity.NetworkInterfaceBondParams {
+
+	return &entity.NetworkInterfaceBondParams{
+		NetworkInterfacePhysicalParams: entity.NetworkInterfacePhysicalParams{
+			MACAddress: d.Get("mac_address").(string),
+			Name:       d.Get("name").(string),
+			Tags:       strings.Join(convertToStringSlice(d.Get("tags").(*schema.Set).List()), ","),
+			VLAN:       d.Get("vlan").(string),
+			MTU:        d.Get("mtu").(int),
+			AcceptRA:   d.Get("accept_ra").(bool),
+			Autoconf:   d.Get("autoconf").(bool),
+		},
+		Parents:            parentIDs,
+		BondMode:           d.Get("bond_mode").(string),
+		BondMiimon:         d.Get("bond_miimon").(int),
+		BondDownDelay:      d.Get("bond_downdelay").(int),
+		BondUpDelay:        d.Get("bond_updelay").(int),
+		BondLACPRate:       d.Get("bond_lacp_rate").(string),
+		BondXMitHashPolicy: d.Get("bond_xmit_hash_policy").(string),
+		BondNumberGratARP:  d.Get("bond_num_grat_arp").(int),
+	}
```

Comment:
most of these values can be `nil` and type asserted to a type where `nil` is invalid, those type assertions should be checked.

Reply:
I dont believe it will be nil.  The key is defined in the schema.
- https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#ResourceData.Get

> Get returns the data for the given key, or nil if the key doesn't exist in the schema.
>
>If the key does exist in the schema but doesn't exist in the configuration, then the default value for that type will be >returned. For strings, this is "", for numbers it is 0, etc.
>
>If you want to test if something is set at all in the configuration, use GetOk.